internal/api_server/agentserver/grpc.go

```diff
@@ -108,43 +132,116 @@ func (s *AgentGrpcServer) Stream(stream pb.RouterService_StreamServer) error {
-		if actual.(streamCtx).closed || otherSideStream == nil {
+		otherSideSession := actual.(streamCtx).session
+		if actual.(streamCtx).closed || otherSideStream == nil && otherSideSession == nil {
 			// the other side closed the connection, we should not accept any more messages
 			s.log.Infof("client %s, attempted connection to %s which was already closed", clientName, sessionId)
 			return nil
 		}
-		err := forward(ctx, stream, otherSideStream)
-		if errors.Is(err, io.EOF) {
-			// one side closed the connection, we should not accept any more messages
-			s.log.Infof("one client disconnected from session %s, closing", sessionId)
-
-			// we try to send a close message to both sides, no error checking, best effort
-			err := stream.Send(&pb.StreamResponse{Closed: true})
-			if err != nil {
-				s.log.Warningf("sending close message to stream %s: %s", sessionId, err)
-			}
-
-			err = otherSideStream.Send(&pb.StreamResponse{Closed: true})
-			if err != nil {
-				s.log.Warningf("sending close message to stream %s: %s", sessionId, err)
-			}
-
-			// free the other stream reference, and mark this one as closed
-			stctx := actual.(streamCtx)
-			stctx.stream = nil
-			stctx.closed = true
-			s.pendingStreams.Store(sessionId, stctx)
-			return nil
-		} else {
-			actual.(streamCtx).cancel()
+		// TODO(majopela): handling of the gRPC<->gRPC communication, we should remove this once the console UI is fully migrated to the new ws API
+		if otherSideStream != nil {
+			s.log.Infof("client %s connected to session %s, forwarding gRPC<->gRPC", clientName, sessionId)
+			err := s.forwardGrpcOnlyStream(ctx, stream, otherSideStream, sessionId, actual)
+			return err
+		}
+		// There is a ws console session on the other side, we should forward the ws connection to the console session
+		if otherSideSession != nil {
+			s.log.Infof("client %s connected to session %s, forwarding ws console session", clientName, sessionId)
+			err := s.forwardConsoleSession(ctx, stream, otherSideSession, sessionId, actual)
 			return err
 		}
+
 	} else {
 		// the map did not have a value, we are the first client, so we wait for the second
 		s.log.Infof("client %s waiting for peer %s", clientName, sessionId)
 		<-ctx.Done()
 		return nil
 	}
+	return nil
+}
+
+func (s *AgentGrpcServer) forwardConsoleSession(ctx context.Context, stream pb.RouterService_StreamServer, otherSideSession *console.ConsoleSession, sessionId string, actual any) error {
+	err := s.forwardChannels(ctx, stream, otherSideSession)
+
+	return err
+
+}
+
+func (s *AgentGrpcServer) forwardChannels(ctx context.Context, a pb.RouterService_StreamServer, b *console.ConsoleSession) error {
+	g, _ := errgroup.WithContext(ctx)
+	g.Go(func() error { return s.pipeStreamToChannel(a, b.RecvCh) })
+	g.Go(func() error { return s.pipeChannelToStream(b.SendCh, a) })
+	return g.Wait()
+}
+
+func (s *AgentGrpcServer) pipeStreamToChannel(a pb.RouterService_StreamServer, ch chan []byte) error {
```

Comment:
can we add a ctx this this signature and add a select case for <-ctx.Done()?

Reply:
Done, it's not perfect because a.Recv is blocking, but as soon as it loops once it will find the context Done and get out.